endorsedLibraries/kotlinx.cli/src/main/kotlin/kotlinx/cli/Descriptors.kt

```diff
@@ -0,0 +1,152 @@
+ *
+ * @property type option/argument type, one of [ArgType].
+ * @property fullName option/argument full name.
+ * @property description text description of option/argument.
+ * @property defaultValue default value for option/argument.
+ * @property required if option/argument is required or not. If it's required and not provided in command line and have no default value, error will be generated.
+ * @property deprecatedWarning text message with information in case if option is deprecated.
+ */
+internal abstract class Descriptor<T : Any, TResult>(val type: ArgType<T>,
+                                                     var fullName: String? = null,
+                                                     val description: String? = null,
+                                                     val defaultValue: TResult? = null,
+                                                     val required: Boolean = false,
+                                                     val deprecatedWarning: String? = null) {
+    /**
+     * Text description for help message.
+     */
+    abstract val textDescription: String
+    /**
+     * Help message for descriptor.
+     */
+    abstract val helpMessage: String
+
+    /**
+     * Provide text description of value.
+     *
+     * @param value value got getting text description for.
+     */
+    fun valueDescription(value: TResult?) = value?.let {
+        if (it is List<*> && it.isNotEmpty())
+            " [${it.joinToString { it.toString() }}]"
+        else if (it !is List<*>)
+            " [$it]"
+        else null
+    }
+
+    /**
+     * Flag to check if descriptor has set default value for option/argument.
+     */
+    val defaultValueSet by lazy {
+        defaultValue != null && (defaultValue is List<*> && defaultValue.isNotEmpty() || defaultValue !is List<*>)
+    }
+}
+
+/**
+ * Option descriptor.
+ *
+ * Command line entity started with some prefix (-/--) and can have value as next entity in command line string.
+ *
+ * @property optionFullFormPrefix prefix used before full form of option.
+ * @property optionShortFromPrefix prefix used before short form of option.
+ * @property type option type, one of [ArgType].
+ * @property fullName option full name.
+ * @property shortName option short name.
+ * @property description text description of option.
+ * @property defaultValue default value for option.
+ * @property required if option is required or not. If it's required and not provided in command line and have no default value, error will be generated.
+ * @property multiple if option can be repeated several times in command line with different values. All values are stored.
+ * @property delimiter delimiter that separate option provided as one string to several values.
+ * @property deprecatedWarning text message with information in case if option is deprecated.
+ */
+internal class OptionDescriptor<T : Any, TResult>(
+        val optionFullFormPrefix: String,
```

Comment:
Have to add prefix form as property for options, because of making modifiers as open part of API, so `OptionDescriptor` can't be inner class any more.